src/ext/oblivc/ot.c

```diff
@@ -1033,6 +1080,69 @@ bcipherCryptNoResize(BCipherRandomGen* gen,const char* key,int nonce,
   randomizeBuffer(gen,dest,n);
   memxor(dest,src,n);
 }
+
+
+/* 
+   Data structures and functions for dealing with multithreaded nonce-allocation
+   We use gcc's built-in atomics rather than synchronization primitives. On modern CPUs
+   these translate directly into instructions, so they're pretty quick. The reason for
+   all this complication: setting up OT extensions afresh is really expensive, so we
+   want to be able to reuse a single instance across multiple threads
+   */
+
+typedef struct HonestOTExtNonceTracker
+{ size_t * nonce_master;
+  size_t * ref_count;
+} HonestOTExtNonceTracker;
+
+typedef struct HonestOTExtSender
+{ SenderExtensionBox* box;
+  BCipherRandomGen* padder;
+  HonestOTExtNonceTracker* nonce_trk;
+} HonestOTExtSender;
+
+typedef struct HonestOTExtRecver
+{ RecverExtensionBox* box;
+  BCipherRandomGen* padder;
+} HonestOTExtRecver;
+
+void honestOTExtNonceTrackerInit(HonestOTExtNonceTracker * t, HonestOTExtNonceTracker * t_prev)
+{ if (t_prev == NULL)
+  {
+    t->nonce_master = malloc(sizeof(*t->nonce_master));
+    t->ref_count = malloc(sizeof(*t->ref_count));
+    *t->ref_count = 1;
+  }
+  else
+  {
+    t->nonce_master = t_prev->nonce_master;
+    t->ref_count = t_prev->ref_count;
+    __atomic_add_fetch(t->ref_count, 1, __ATOMIC_RELAXED);
```

Comment:
I don't remember if I've already said this. Can we try C11 atomics instead of GNU extensions? Or do they not work for some reason?

Reply:
We can use C11 atomics if you like, but it's worth noting that they're only supported for GCC >= 4.9, whereas the GCC builtins go back to 4.1. Do you care about support for those versions?